File: learning-factory/warehouse/artifacts/projects/build-your-own-x/programming-language/a-compiler-from-scratch-84427e53fe/job_byox_build_v1_f3e0d99cf97b485202190e3be4a0e8f6/attempt-001/sealed/reference_tests/validate_structure.rb
# frozen_string_literal: true

require "json"
require "find"

REQUIRED = [
  "README.md", "AGENTS.md", "MANIFEST.yaml", "PROVENANCE.json",
  "LICENSE_BOUNDARY.md", "REQUIREMENTS.md", "CONCEPTS.md",
  "DESIGN_QUESTIONS.md", "VALIDATION.md", "starter/README.md",
  "public_tests/README.md", "environment/README.md",
  "sealed/reference/README.md", "sealed/reference_tests/README.md",
  "sealed/DESIGN.md", "sealed/TRADEOFFS.md", "sealed/REVIEW.md",
  "sealed/alternatives/README.md",
  "sealed/production/PRODUCTIONIZATION.md", "adversarial/README.md",
  "debugging/README.md", "review_exercises/README.md",
  "benchmarks/README.md"
].freeze

FORBIDDEN = [
  ".git", ".env", ".venv", "credentials.json", "secrets", "reference",
  "reference_tests", "hidden_tests", "solution", "solutions", "answers",
  "starter/sealed", "starter/reference", "starter/reference_tests",
  "starter/solution", "starter/solutions", "starter/answers",
  "public_tests/sealed", "public_tests/reference", "public_tests/hidden_tests",
  "environment/sealed"
].freeze

GENERATED_ROOTS = [
  "README.md", "AGENTS.md", "MANIFEST.yaml", "PROVENANCE.json",
  "LICENSE_BOUNDARY.md", "REQUIREMENTS.md", "CONCEPTS.md",
  "DESIGN_QUESTIONS.md", "VALIDATION.md", "starter", "public_tests",
  "environment", "sealed", "adversarial", "debugging", "review_exercises",
  "benchmarks"
].freeze

MANIFEST = {
  "independent_validation" => "REQUIRED",
  "productionized" => false,
  "project_id" => "project_0d336967c5b89e5c4851b06a9e793cae",
  "provenance_sha256" => "e534f25088ceaa9ff361d1831402338ec743bc666427d4b05cb4f56be11ea594",
  "schema_version" => 1,
  "source_commit" => "aa17439b62f384511a5561ce308e9598b94d8989",
  "source_id" => "source_eac489a34bed5db9a1f2a580b457bcef",
  "status" => "GENERATED",
  "validation_labels" => ["GENERATED", "PARTIAL"]
}.freeze

missing = REQUIRED.reject { |path| File.file?(path) }
present_forbidden = FORBIDDEN.select { |path| File.exist?(path) || File.symlink?(path) }
raise "missing required: #{missing.inspect}" unless missing.empty?
raise "forbidden paths: #{present_forbidden.inspect}" unless present_forbidden.empty?

manifest = JSON.parse(File.read("MANIFEST.yaml"))
raise "manifest mismatch" unless manifest == MANIFEST

provenance = JSON.parse(File.read("PROVENANCE.json"))
expected_top = %w[classification license_boundary project schema_version snapshot_sha256 source]
raise "provenance top-level mismatch" unless provenance.keys.sort == expected_top.sort
raise "provenance snapshot mismatch" unless provenance["snapshot_sha256"] == manifest["provenance_sha256"]
raise "provenance project mismatch" unless provenance["project"]["project_id"] == manifest["project_id"]
raise "provenance source mismatch" unless provenance["source"]["source_id"] == manifest["source_id"]
unless provenance["license_boundary"]["linked_content_copied"] == false &&
       provenance["license_boundary"]["linked_resource_license"] == "NOASSERTION"
  raise "linked-content license boundary mismatch"
end

special = []
files = []
GENERATED_ROOTS.each do |root|
  Find.find(root) do |path|
    stat = File.lstat(path)
    if stat.symlink?
      special << path
      Find.prune if File.directory?(path)
    elsif stat.directory?
      next
    elsif stat.file?
      files << path
    else
      special << path
    end
  end
end
raise "non-regular generated entries: #{special.inspect}" unless special.empty?

credential_patterns = [
  /-----BEGIN (?:RSA |EC |DSA |OPENSSH )?PRIVATE KEY-----/,
  /\b(?:api[_-]?key|access[_-]?token|auth[_-]?token|client[_-]?secret|password)\s*[:=]\s*["'][^"'\r\n]+/i,
  /\bAKIA[0-9A-Z]{16}\b/,
  /\bgh[pousr]_[A-Za-z0-9]{20,}\b/,
  /\bxox[baprs]-[A-Za-z0-9-]{10,}\b/
]
credential_hits = []
files.each do |path|
  content = File.binread(path)
  credential_patterns.each do |pattern|
    credential_hits << [path, pattern.source] if content.match?(pattern)
  end
end
raise "credential-like content: #{credential_hits.inspect}" unless credential_hits.empty?

answer_leaks = files.select do |path|
  File.basename(path).match?(/ANSWER|SOLUTION/i) && !path.start_with?("sealed/")
end
raise "solution-bearing filename outside sealed: #{answer_leaks.inspect}" unless answer_leaks.empty?

puts "required files: #{REQUIRED.length}/#{REQUIRED.length}"
puts "forbidden paths present: #{present_forbidden.length}"
puts "non-regular generated entries: #{special.length}"
puts "credential-pattern matches: #{credential_hits.length}"
puts "solution-bearing filenames outside sealed: #{answer_leaks.length}"
puts "manifest strict object: OK"
puts "provenance strict JSON and identifiers: OK"
puts "generated regular files scanned: #{files.length}"

File: learning-factory/warehouse/artifacts/projects/build-your-own-x/remediation/e2688c6c956a61b0a4f6/baseline-9f7fe579458899db6a5e/repair-v1/job_byox_repair_s2_v1_g1_4f89d5c495e9764a76bbc61a5b688d33/attempt-001/environment/validate_structure.py
"""Deterministic archive-safety and metadata checks for this challenge pack."""

from __future__ import annotations

import json
import os
import re
import stat
import sys
from pathlib import Path


ROOT = Path(__file__).resolve().parent.parent

REQUIRED = (
    "README.md",
    "AGENTS.md",
    "MANIFEST.yaml",
    "PROVENANCE.json",
    "LICENSE_BOUNDARY.md",
    "REQUIREMENTS.md",
    "CONCEPTS.md",
    "DESIGN_QUESTIONS.md",
    "VALIDATION.md",
    "starter/README.md",
    "public_tests/README.md",
    "environment/README.md",
    "sealed/reference/README.md",
    "sealed/reference_tests/README.md",
    "sealed/DESIGN.md",
    "sealed/TRADEOFFS.md",
    "sealed/REVIEW.md",
    "sealed/alternatives/README.md",
    "sealed/production/PRODUCTIONIZATION.md",
    "adversarial/README.md",
    "debugging/README.md",
    "review_exercises/README.md",
    "benchmarks/README.md",
)

FORBIDDEN = (
    ".git",
    ".env",
    ".venv",
    "credentials.json",
    "secrets",
    "reference",
    "reference_tests",
    "hidden_tests",
    "solution",
    "solutions",
    "answers",
    "starter/sealed",
    "starter/reference",
    "starter/reference_tests",
    "starter/solution",
    "starter/solutions",
    "starter/answers",
    "public_tests/sealed",
    "public_tests/reference",
    "public_tests/hidden_tests",
    "environment/sealed",
)

EXPECTED_MANIFEST = {
    "independent_validation": "REQUIRED",
    "productionized": False,
    "project_id": "project_ce9cae8ab2c9c820ceec425c2639e501",
    "provenance_sha256":
        "7cedc44e4fb22194236a8939bd566c5734dfdb3d379f192ed8c7ce1e7159ef6a",
    "schema_version": 1,
    "source_commit": "aa17439b62f384511a5561ce308e9598b94d8989",
    "source_id": "source_eac489a34bed5db9a1f2a580b457bcef",
    "status": "GENERATED",
    "validation_labels": ["GENERATED", "PARTIAL"],
}

SECRET_PATTERNS = (
    re.compile(rb"-----BEGIN (?:RSA |EC |OPENSSH )?PRIVATE KEY-----"),
    re.compile(rb"AKIA[0-9A-Z]{16}"),
    re.compile(rb"ghp_[A-Za-z0-9]{30,}"),
    re.compile(rb"sk-[A-Za-z0-9]{32,}"),
)

# Workspace-control directories are owned by the harness, not generated pack
# material. Do not descend into or archive-scan them.
CONTROL_DIRECTORIES = {".agents", ".codex", ".factory-workspace"}


def main() -> int:
    errors: list[str] = []

    for relative in REQUIRED:
        path = ROOT / relative
        if not path.is_file():
            errors.append(f"missing required regular file: {relative}")

    for relative in FORBIDDEN:
        path = ROOT / relative
        if path.exists() or path.is_symlink():
            errors.append(f"forbidden path exists: {relative}")

    for current, directory_names, file_names in os.walk(ROOT, followlinks=False):
        current_path = Path(current)
        if current_path == ROOT:
            directory_names[:] = [
                name for name in directory_names
                if name not in CONTROL_DIRECTORIES
            ]
        for name in directory_names + file_names:
            path = current_path / name
            relative = path.relative_to(ROOT).as_posix()
            mode = path.lstat().st_mode
            if stat.S_ISLNK(mode):
                errors.append(f"symbolic link is not archivable: {relative}")
            elif not (stat.S_ISREG(mode) or stat.S_ISDIR(mode)):
                errors.append(f"special filesystem object: {relative}")
            elif stat.S_ISREG(mode):
                content = path.read_bytes()
                for pattern in SECRET_PATTERNS:
                    if pattern.search(content):
                        errors.append(f"credential-like material in: {relative}")
                        break

    try:
        manifest = json.loads((ROOT / "MANIFEST.yaml").read_text("utf-8"))
    except (OSError, UnicodeError, json.JSONDecodeError) as error:
        errors.append(f"MANIFEST.yaml is not strict JSON: {error}")
    else:
        if manifest != EXPECTED_MANIFEST:
            errors.append("MANIFEST.yaml does not equal the required object")

    try:
        provenance = json.loads((ROOT / "PROVENANCE.json").read_text("utf-8"))
    except (OSError, UnicodeError, json.JSONDecodeError) as error:
        errors.append(f"PROVENANCE.json is not strict JSON: {error}")
    else:
        if provenance.get("snapshot_sha256") != EXPECTED_MANIFEST["provenance_sha256"]:
            errors.append("provenance snapshot does not match the manifest")

    if errors:
        for error in errors:
            print(f"ERROR: {error}")
        return 1
    print(f"structure: PASS ({len(REQUIRED)} required files, no forbidden paths)")
    print("archive types: PASS (regular files and directories only)")
    print("credential patterns: PASS")
    print("metadata status: PASS (GENERATED + PARTIAL)")
    return 0


if __name__ == "__main__":
    sys.exit(main())
